### PR TITLE
update type name accesses

### PR DESCRIPTION
**Pull request by pq**
API updates.

/cc @scheglov @bwilkerson

### CODE REVIEW

File: lib/src/rules/prefer_int_literals.dart [hunk 1]
```diff
@@ -71,26 +71,26 @@ class _Visitor extends SimpleAstVisitor<void> {
...
   bool hasTypeDouble(Expression expression) {
     final parent = expression.parent;
     if (parent is ArgumentList) {
-      return expression.staticParameterElement?.type?.name == 'double';
+      return expression.staticParameterElement?.type?.isDartCoreDouble == true;
     } else if (parent is ListLiteral) {
       final typeArguments = parent.typeArguments?.arguments;
       return typeArguments?.length == 1 &&
-          typeArguments[0]?.type?.name == 'double';
+          typeArguments[0]?.type?.isDartCoreDouble == true;
```
Comment (scheglov) [inline lib/src/rules/prefer_int_literals.dart]: `.type` of a type arguments cannot be null, so could be just `typeArguments[0].type.isDartCoreDouble`

File: lib/src/rules/prefer_int_literals.dart [hunk 2]
```diff
@@ -71,26 +71,26 @@ class _Visitor extends SimpleAstVisitor<void> {
...
     } else if (parent is NamedExpression) {
```
Comment (scheglov) [inline lib/src/rules/prefer_int_literals.dart]: Maybe call `hasTypeDouble(parent.parent)`

File: lib/src/rules/unrelated_type_equality_checks.dart [hunk 3]
```diff
@@ -146,11 +144,10 @@ bool _hasNonComparableOperands(TypeSystem typeSystem, BinaryExpression node) {
       !(_isFixNumIntX(leftType) && _isCoreInt(rightType));
 }
 
-bool _isCoreInt(DartType type) =>
-    type.name == 'int' && type.element?.library?.name == _dartCoreLibraryName;
+bool _isCoreInt(DartType type) => type.isDartCoreInt;
 
 bool _isFixNumIntX(DartType type) =>
-    (type.name == 'Int32' || type.name == 'Int64') &&
+    (type.element?.name == 'Int32' || type.element?.name == 'Int64') &&
```
Comment (scheglov) [inline lib/src/rules/unrelated_type_equality_checks.dart]: I'd replaced this with `type is InterfaceType`, and then call `type.element.name` without `null` protection.
Maybe extract `element` into a local variable.

File: lib/src/rules/use_full_hex_values_for_flutter_colors.dart [hunk 4]
```diff
@@ -54,7 +54,7 @@ class _Visitor extends SimpleAstVisitor {
   @override
   void visitInstanceCreationExpression(InstanceCreationExpression node) {
     if (node.staticElement?.library?.name == 'dart.ui' &&
-        node.staticElement?.returnType?.name == 'Color' &&
+        node.staticElement?.returnType?.element?.name == 'Color' &&
         node.staticElement?.name == '') {
```
Comment (scheglov) [inline lib/src/rules/use_full_hex_values_for_flutter_colors.dart]: I'd replaced this code with a call to something more generic, like `isConstructorElement(ConstructorElement element, {@required String uriStr, @required String className, @required String constructorName})`.

It is better to check `ConstructorElement.enclosingElement`, which is `ClassElement` than `returnType.element`. Although we probably should override `ConstructorElement.returnType` to return `InterfaceType`.

File: lib/src/util/dart_type_utilities.dart [hunk 5]
```diff
@@ -176,12 +176,12 @@ class DartTypeUtilities {
 
   static bool isClass(DartType type, String className, String library) =>
       type != null &&
```
Comment (scheglov) [inline lib/src/util/dart_type_utilities.dart]: `is InterfaceType`, and then unprotected accesses.

File: lib/src/util/dart_type_utilities.dart [hunk 6]
```diff
@@ -176,12 +176,12 @@ class DartTypeUtilities {
...
-      type.name == className &&
+      type.element?.name == className &&
       type.element?.library?.name == library;
 
   static bool isInterface(
           InterfaceType type, String interface, String library) =>
-      type.name == interface && type.element.library.name == library;
+      type.element?.name == interface && type.element.library.name == library;
```
Comment (scheglov) [inline lib/src/util/dart_type_utilities.dart]: `type.element` cannot be `null`

File: lib/src/util/dart_type_utilities.dart [hunk 7]
```diff
@@ -176,12 +176,12 @@ class DartTypeUtilities {
 
   static bool isClass(DartType type, String className, String library) =>
       type != null &&
-      type.name == className &&
+      type.element?.name == className &&
       type.element?.library?.name == library;
 
   static bool isInterface(
```
Comment (scheglov) [inline lib/src/util/dart_type_utilities.dart]: I see very little difference between this and the previous function.
I'd leave one, called `isInterfaceType()`.

File: lib/src/rules/null_closures.dart [hunk 8]
```diff
@@ -313,7 +313,8 @@ class _Visitor extends SimpleAstVisitor<void> {
       return null;
     }
     for (var supertype in element.allSupertypes) {
-      method = getMethod(supertype.element.library.name, supertype.name);
+      method =
+          getMethod(supertype.element.library.name, supertype.element?.name);
```
[the remark this answers is not in the document]
Comment (pq) [reply lib/src/rules/null_closures.dart]: Great.  Thanks!

File: lib/src/rules/prefer_int_literals.dart (hunk 1 above)
Comment (pq) [reply lib/src/rules/prefer_int_literals.dart]: 👍

File: lib/src/rules/unrelated_type_equality_checks.dart (hunk 3 above)
Comment (pq) [reply lib/src/rules/unrelated_type_equality_checks.dart]: 👍

File: lib/src/util/dart_type_utilities.dart (hunk 5 above)
Comment (pq) [reply lib/src/util/dart_type_utilities.dart]: 👍

File: lib/src/util/dart_type_utilities.dart (hunk 7 above)
Comment (pq) [reply lib/src/util/dart_type_utilities.dart]: Good catch.

File: lib/src/rules/prefer_int_literals.dart [hunk 9]
```diff
@@ -71,26 +71,26 @@ class _Visitor extends SimpleAstVisitor<void> {
     if (node is FunctionExpression) {
       var functDeclaration = node.parent;
       if (functDeclaration is FunctionDeclaration) {
-        return functDeclaration.returnType?.type?.name == 'double';
+        return functDeclaration.returnType?.type?.isDartCoreDouble == true;
       }
     } else if (node is MethodDeclaration) {
-      return node.returnType?.type?.name == 'double';
+      return node.returnType?.type?.isDartCoreDouble == true;
```
[the remark this answers is not in the document]
Comment (pq) [reply lib/src/rules/prefer_int_literals.dart]: Agreed.

File: lib/src/rules/use_full_hex_values_for_flutter_colors.dart (hunk 4 above)
Comment (pq) [reply lib/src/rules/use_full_hex_values_for_flutter_colors.dart]: Good idea(s).  Updated.  Thanks!

File: lib/src/rules/null_closures.dart [hunk 10]
```diff
@@ -303,7 +303,7 @@ class _Visitor extends SimpleAstVisitor<void> {
         possibleMethods
             .lookup(NonNullableFunction(library, className, methodName));
 
-    var method = getMethod(type.element.library.name, type.name);
+    var method = getMethod(type.element.library.name, type.element?.name);
```
[the remark this answers is not in the document]
Comment (pq) [reply lib/src/rules/null_closures.dart]: Yes.  That was overprotective.

Thanks.

**Pull request merged**